### ISSUE TITLE
Zsys changes the canmount property of persistent bpool/grub dataset 

### ISSUE THREAD

**Issue by ddnexus** [ubuntu/zsys#164]
As partially discussed in #163 with @rlaager I can confirm that if you create a persistent set for grub (especially needed for mirror setups) like the following:
```
zfs create -o canmount=on -o mountpoint=/boot/grub bpool/grub
```
and update grub, after the first reboot it will continue to be `canmount=on`, but after another (with a restore) you will get `canmount=noauto` and it will not get mounted from then on.

One consequence is that `/boot/grub` seems to be written in the /boot dataset and contains only `grub.cfg` and `grubenv`. You can `rm -rf /boot/grub` restore `canmount=on` and check that is mounted, upgrade grub, and after the same 2 reboots it will be in the same state (i.e. canmount=noauto, not mounted and `/boot/grub` containing the 2 mentioned files only).

Not sure whether it is triggered by the mount point being `/boot/grub` or by the `bpool/grub` dataset name, but there is a confirmed problem with mounting it.

That could also be (part of or related to) the problem of the grub menu being the same after a restore: it's like the actual `grub.cfg` that should be mounted from the dataset, it's not available/mounted at boot time, so grub gets the same file already in the `/boot` dataset.

**Comment by didrocks** [comment]
I was able to reproduce the issue with the minimal reproducer bug:
- install with default ZSys/ZFS option on ubuntu
- then migrate it to the newly created dataset by:
  - remove **/boot/efi/grub  /boot/grub      none    defaults,bind   0       0** from `/etc/fstab`
  - `sudo umount /boot/grub`
  - `sudo zfs create -o canmount=on -o mountpoint=/boot/grub bpool/grub`
  - `sudo cp -a /boot/efi/grub/*  /boot/grub`
  - `sudo update-grub && sudo grub-install /dev/vda`
-> /boot/grub/grub.cfg is updated and zfs mount shows the correct one is mounted:
```
$ zfs mount
[…]
bpool/grub                      /boot/grub
```

After a reboot persistent grub is still mounted:
```
$ zfs mount
[…]
bpool/grub                      /boot/grub
```

- taking a snapshot of the new layout:
```
$ sudo zsysctl save --system
ZSys is adding automatic system snapshot to GRUB menu
Successfully saved as "autozsys_zz0e21"
```
-> /boot/grub/grub.cfg is updated with the new snapshot.
```
$ grep autozsys_zz0e21 /boot/grub/grub.cfg 
	submenu 'Revert to 21/09/2020 @ 11:18' ${menuentry_id_option} 'gnulinux-history-rpool/ROOT/ubuntu_qiq15o@autozsys_zz0e21' {
		zsyshistorymenu "rpool/ROOT/ubuntu_qiq15o@autozsys_zz0e21" "/dev/vda6" "/BOOT/ubuntu_qiq15o@autozsys_zz0e21/initrd.img-5.4.0-47-generic" "/BOOT/ubuntu_qiq15o@autozsys_zz0e21/vmlinuz-5.4.0-47-generic" "5.4.0-47-generic"
```

- Rebooting and reverting to the snapshot with this layout, shows, indeed that there is no bpool/grub and canmount was set to noauto:
```
$ zfs get canmount bpool/grub 
NAME        PROPERTY  VALUE     SOURCE
bpool/grub  canmount  noauto    local
```

Thanks for the report and indeed, that shouldn’t happen, I’ll investigate this :)

**Comment by ddnexus** [comment]
Thank you for confirming it so quickly!

I am going to abandon the manual setup for mirror that does not work because of this bug, and reinstall it with the Ubuntu installer, clone the partitions on the other SSD and attach the new partitions to the existing pools.

The problem with grub in ESP with mirror is that you have to keep both ESPs grub in sync and the fstab cannot mount  `/boot/grub` from one or the other conditionally, hence using a persistent dataset canmount=on would solve the problem.

What should we expect to be the setup of grub for mirror when it will be supported?

**Comment by didrocks** [comment]
I had a look this morning and the above PR should fix it. If you want to test it (there are new tests to cover it, as you can see in the PR, so I’m quite confident), I can provide you a binary :)

**Comment by ddnexus** [comment]
@didrocks you rock! :)

I will be able to test it on my (still delayed) new installation probably tomorrow when I will have time!

Thank you!!!

**Comment by ddnexus** [comment]
And of course, I will need some instruction/binary or whatever in order to use the changes. Thank you!

**Comment by rlaager** [comment]
@didrocks Any chance this can get backported into 20.04?

**Comment by ddnexus** [comment]
@didrocks what will be the zsys version that will include the fix?

**Comment by didrocks** [comment]
The fix is available for groovy (to become 20.10) as https://launchpad.net/ubuntu/+source/zsys/0.5.2.

I have backported the fix and uploaded to 20.04 LTS. Once the SRU team accepts it (which can take some days), it will be available in -proposed and you can track, help testing looking at the bug https://bugs.launchpad.net/ubuntu/+source/zsys/+bug/1897563 (which will be updated). The version in 20.04 LTS for ZSys will be 0.4.8.
